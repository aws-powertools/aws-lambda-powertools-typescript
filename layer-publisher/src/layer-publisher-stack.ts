import { CfnOutput, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
  LayerVersion,
  Code,
  Runtime,
  CfnLayerVersionPermission
} from 'aws-cdk-lib/aws-lambda';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';

export interface LayerPublisherStackProps extends StackProps {
  readonly layerName?: string
  readonly powerToolsPackageVersion?: string
  readonly ssmParameterLayerArn: string
}

export class LayerPublisherStack extends Stack {
  public readonly lambdaLayerVersion: LayerVersion;
  public constructor(scope: Construct, id: string, props: LayerPublisherStackProps) {
    super(scope, id, props);

    const { layerName, powerToolsPackageVersion } = props;

    console.log(`publishing layer ${layerName} version : ${powerToolsPackageVersion}`);

    this.lambdaLayerVersion = new LayerVersion(this, 'LambdaPowertoolsLayer', {
      layerVersionName: props?.layerName,
      description: `AWS Lambda Powertools for TypeScript version ${powerToolsPackageVersion}`,
      compatibleRuntimes: [
        Runtime.NODEJS_14_X,
        Runtime.NODEJS_16_X,
        Runtime.NODEJS_18_X
      ],
      code: Code.fromAsset('../tmp'),
    });

    const layerPermission = new CfnLayerVersionPermission(this, 'PublicLayerAccess', {
      action: 'lambda:GetLayerVersion',
      layerVersionArn: this.lambdaLayerVersion.layerVersionArn,
      principal: '*',
    });

    layerPermission.applyRemovalPolicy(RemovalPolicy.RETAIN);
    this.lambdaLayerVersion.applyRemovalPolicy(RemovalPolicy.RETAIN);

    new StringParameter(this, 'VersionArn', {
      parameterName: props.ssmParameterLayerArn,
      stringValue: this.lambdaLayerVersion.layerVersionArn,
    });
    new CfnOutput(this, 'LatestLayerArn', {
      value: this.lambdaLayerVersion.layerVersionArn,
      exportName: props?.layerName ?? `LambdaPowerToolsForTypeScriptLayerARN`,
    });
  }
}
