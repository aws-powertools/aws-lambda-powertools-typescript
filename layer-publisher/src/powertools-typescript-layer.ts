import { LayerVersion, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { Md5 } from 'ts-md5';

export interface PowertoolsTypeScriptLayerProps {
  /**
   * The powertools package version from npm repository.
   */
  readonly version?: string

  /**
   * the name of the layer, will be randomised by cdk if empty
   */
  readonly layerVersionName?: string
}

export class PowertoolsTypeScriptLayer extends LayerVersion {
  public constructor(scope: Construct, id: string, props?: PowertoolsTypeScriptLayerProps) {
    const version = props?.version ?? 'latest';
    console.log(`publishing layer ${props?.layerVersionName} version : ${version}`);

    super(scope, id, {
      layerVersionName: props?.layerVersionName,
      description: `AWS Lambda Powertools for TypeScript version ${props?.version}`,
      compatibleRuntimes: [
        Runtime.NODEJS_14_X,
        Runtime.NODEJS_16_X,
        Runtime.NODEJS_18_X
      ],
      code: Code.fromAsset('../tmp', {
        assetHash: Md5.hashStr(version),
        bundling: {
          image: Runtime.NODEJS_14_X.bundlingImage,
        },
      }),
    });
  }
}
