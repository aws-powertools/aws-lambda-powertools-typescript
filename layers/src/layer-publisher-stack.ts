import { CfnOutput, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
  LayerVersion,
  Code,
  Runtime,
  CfnLayerVersionPermission,
} from 'aws-cdk-lib/aws-lambda';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';

export interface LayerPublisherStackProps extends StackProps {
  readonly layerName?: string;
  readonly powertoolsPackageVersion?: string;
  readonly ssmParameterLayerArn: string;
}

export class LayerPublisherStack extends Stack {
  public readonly lambdaLayerVersion: LayerVersion;
  public constructor(
    scope: Construct,
    id: string,
    props: LayerPublisherStackProps
  ) {
    super(scope, id, props);

    const { layerName, powertoolsPackageVersion } = props;

    console.log(
      `publishing layer ${layerName} version : ${powertoolsPackageVersion}`
    );

    this.lambdaLayerVersion = new LayerVersion(this, 'LambdaPowertoolsLayer', {
      layerVersionName: props?.layerName,
      description: `Powertools for AWS Lambda (TypeScript) version ${powertoolsPackageVersion}`,
      compatibleRuntimes: [
        Runtime.NODEJS_14_X,
        Runtime.NODEJS_16_X,
        Runtime.NODEJS_18_X,
      ],
      license: 'MIT-0',
      // This is needed because the following regions do not support the compatibleArchitectures property #1400
      // ...(![ 'eu-south-2', 'eu-central-2', 'ap-southeast-4' ].includes(Stack.of(this).region) ? { compatibleArchitectures: [Architecture.X86_64] } : {}),
      code: Code.fromAsset('../tmp'),
    });

    const layerPermission = new CfnLayerVersionPermission(
      this,
      'PublicLayerAccess',
      {
        action: 'lambda:GetLayerVersion',
        layerVersionArn: this.lambdaLayerVersion.layerVersionArn,
        principal: '*',
      }
    );

    layerPermission.applyRemovalPolicy(RemovalPolicy.RETAIN);
    this.lambdaLayerVersion.applyRemovalPolicy(RemovalPolicy.RETAIN);

    new StringParameter(this, 'VersionArn', {
      parameterName: props.ssmParameterLayerArn,
      stringValue: this.lambdaLayerVersion.layerVersionArn,
    });
    new CfnOutput(this, 'LatestLayerArn', {
      value: this.lambdaLayerVersion.layerVersionArn,
      exportName: props?.layerName ?? `LambdaPowerToolsForTypeScriptLayerARN`,
    });
  }
}
