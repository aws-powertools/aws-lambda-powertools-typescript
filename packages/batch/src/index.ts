export * from './constants';
export * from './errors';
export * from './types';
export * from './BasePartialProcessor';
export * from './BatchProcessor';
export * from './processPartialResponse';
export * from './SqsFifoPartialProcessor';
