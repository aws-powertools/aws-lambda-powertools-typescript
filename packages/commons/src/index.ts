export * from './utils/lambda';
export * from './utils/middy';
export * as ContextExamples from './tests/resources/contexts';
export * as Events from './tests/resources/events';