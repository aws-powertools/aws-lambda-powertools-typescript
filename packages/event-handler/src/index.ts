export * from './types';
export * from './Exceptions';