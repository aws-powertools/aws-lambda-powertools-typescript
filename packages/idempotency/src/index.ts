export * from './errors';
export * from './IdempotencyConfig';
export * from './makeIdempotent';
export * from './idempotencyDecorator';
export { IdempotencyRecordStatus } from './constants';
