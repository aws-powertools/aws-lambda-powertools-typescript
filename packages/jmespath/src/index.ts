export { search } from './search.js';
export {
  JMESPathError,
  LexerError,
  ParseError,
  IncompleteExpressionError,
  ArityError,
  VariadicArityError,
  JMESPathTypeError,
  EmptyExpressionError,
  UnknownFunctionError,
} from './errors.js';
