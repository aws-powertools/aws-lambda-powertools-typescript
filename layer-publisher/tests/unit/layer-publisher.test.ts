/**
 * Test layer publisher
 *
 * @group unit/layer-publisher
 */

import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as LayerPublisher from '../../src/layer-publisher-stack';

test('Layer  Created', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new LayerPublisher.LayerPublisherStack(app, 'MyTestStack');

  // THEN
  const template = Template.fromStack(stack);

  expect(template).toMatchSnapshot();
});
