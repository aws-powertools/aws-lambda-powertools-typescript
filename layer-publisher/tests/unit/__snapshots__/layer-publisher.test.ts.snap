// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer  Created 1`] = `
{
  "Outputs": {
    "LatestLayerArn": {
      "Export": {
        "Name": "AWSLambdaPowertoolsTypeScript",
      },
      "Value": {
        "Ref": "LambdaPowertoolsLayer733361E9",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaPowertoolsLayer733361E9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "CompatibleRuntimes": [
          "nodejs14.x",
          "nodejs16.x",
          "nodejs18.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "503ad69de28d5c29848b356e48699efb350cedbeeb57ad624e7713b1c063f18b.zip",
        },
        "Description": "Lambda Powertools for TypeScript version 1.0.1",
        "LayerName": "AWSLambdaPowertoolsTypeScript",
      },
      "Type": "AWS::Lambda::LayerVersion",
      "UpdateReplacePolicy": "Retain",
    },
    "PublicLayerAccess": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Action": "lambda:GetLayerVersion",
        "LayerVersionArn": {
          "Ref": "LambdaPowertoolsLayer733361E9",
        },
        "Principal": "*",
      },
      "Type": "AWS::Lambda::LayerVersionPermission",
      "UpdateReplacePolicy": "Retain",
    },
    "VersionArn99E177D3": {
      "Properties": {
        "Name": "/layers/powertools-layer-arn",
        "Type": "String",
        "Value": {
          "Ref": "LambdaPowertoolsLayer733361E9",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
